### PR TITLE
pinctrl: Add pinctrl configs and scripts

### PR DESCRIPTION
**Pull request by nandojve**
This add initial version of pinctrl definitions and scripts. It creates pinctrl definitions from a well defined pinconfig yaml file. This file is filled with all I/O multiplexing and considerations for all Atmel SAM SoCs. The auto generated files should be used inside a <board>-pinctrl.dtsi file to define the pinctrl groups.

In addition this add GH test workflow to evaluate any tests available like scripts/tests.

I'm asking for comments about proposal since this is first step to update SAMs to Zephyr pinctrl API.

### CODE REVIEW

**Review by gmarull — COMMENTED** [review]
@nandojve thanks for this PR! can you split changes into multiple commits to ease review (e.g. autogenerated files should go into their own commit)

File: scripts/sampinctrl.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        variant: Variant information.
+        pins: Pins description.
+
+    Returns:
+        Dictionary with pins configuration.
+    """
+
+    pin_cfgs = []
+
+    pins = OrderedDict(natsorted(pins.items(), key=lambda kv: kv[0]))
```
Comment (gmarull) [inline scripts/sampinctrl.py:153]: just curious: what does `natsorted` do?

File: scripts/sampinctrl.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            continue
+
+        build_atmel_sam_sets(pin_cfgs, pin, pin_cfg["mux"], serie, variant)
+        # How to process extra and system?
+        # if "extra" in pin_cfg.keys():
+        #  build_atmel_sam_sets(pin_cfgs, pin, pin_cfg["extra"],
+        #                       serie, variant)
+        # if "system" in pin_cfg.keys():
+        # build_atmel_sam_sets(pin_cfgs, pin, pin_cfg["system"],
+        #                      serie, variant)
```
[the remark this answers is not in the document]
Comment (nandojve) [reply scripts/sampinctrl.py:166]: yes, Atmel have some intrinsic to select pins for system and extra functions. It is not clear to me if that really necessary. probably use another function variation like ATMEL_SAM_PIO_SYSTEM/ATMEL_SAM_PIO_EXTRA or just add flags to be processed at pinctrl driver. For those, a feedback from @stephanosio,  @mnkp would be very helpful.

File: scripts/sampinctrl.py (hunk 1 above)
Comment (nandojve) [reply scripts/sampinctrl.py:153]: netsorted do true Human Order like p0a, p0b, p1a, p1b, p2a, p2b, p10b, p10d, p20a. So, doesn't matter how pinconfigs files will be defined a real human order is performed at Auto Generated files.

**Comment by nandojve** [comment]
>> @nandojve thanks for this PR! can you split changes into multiple commits to ease review (e.g. autogenerated files should go into their own commit)

> Ok, that is not a problem. I can split before select ready to review.

@gmarull ,

You commented that I should split this PR. The problem is that I need at least a full series SAM or SAM0 to be in to add Zephyr PR side. Otherwise it won't be possible test this changes. This is necessary because when I add pinctrl API on some driver the whole series need to be converted to the new format.

Are you OK in divide PR in only two?

**Comment by gmarull** [comment]
I was not talking about dividing PR, just split into multiple commits. For example, first introduce the script, then its tests, then autogenerated files, etc.

**Comment by nandojve** [comment]
DNM: Checking reverting https://github.com/zephyrproject-rtos/hal_atmel/pull/16#issuecomment-1066159414

**Comment by nandojve** [comment]
> DNM: Checking reverting [#16 (comment)](https://github.com/zephyrproject-rtos/hal_atmel/pull/16#issuecomment-1066159414)

At #43731 I'm building almost everything without errors reverting #16. I would like understand what would be best path once it is necessary fix hal_atmel to move forward with pinctrl. Should I revert in this PR?

**Comment by nandojve** [comment]
Ok, `patch: samd21/r21: Normalize underscore defines` seems to fix underscore problem.

File: scripts/sampinctrl.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+"""
+Utility to autogenerate pinctrl definitions.
+
+Usage::
+    python3 sampinctrl.py [-i /path/to/configs] [-o /path/to/include]
+
+Copyright (c) 2021 Teslabs Engineering S.L.
+Copyright (c) 2022 Gerson Fernando Budke
+SPDX-License-Identifier: Apache 2.0
```
Comment (gmarull) [inline scripts/sampinctrl.py:13]: nit: you can remove this as it's already at the top of this file

File: scripts/sampinctrl.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+"""
+
+import argparse
+from collections import OrderedDict
+from pathlib import Path
+from natsort import natsorted
+import re
+
+import yaml
```
Comment (gmarull) [inline scripts/sampinctrl.py:22]: natsort can be moved together with yaml (it's not from std Python lib)

File: CMakeLists.txt [hunk 5]
```diff
@@ -1 +1,2 @@
+zephyr_include_directories(include)
```
Comment (gmarull) [inline CMakeLists.txt:1]: is this required? (when declaring a dts_root `include` path is automatically added I think)

**Review by gmarull — APPROVED** [review]
LGTM, a few minor comments

File: CMakeLists.txt (hunk 5 above)
Comment (nandojve) [reply CMakeLists.txt:1]: It fails without it.

File: scripts/sampinctrl.py (hunk 3 above)
Comment (nandojve) [reply scripts/sampinctrl.py]: Done!

File: scripts/sampinctrl.py (hunk 4 above)
Comment (nandojve) [reply scripts/sampinctrl.py]: Done!

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: [3.6, 3.7, 3.8, 3.9]
```
Comment (gmarull) [inline .github/workflows/test.yml:11]: suggestion: use 3.7 ... 3.10 (3.6 is EOL)

Comment (nandojve) [reply .github/workflows/test.yml]: Done!

**Comment by gmarull** [comment]
I see the scancode is failing because of invalid spdx tag: `Apache 2.0`, should be `Apache-2.0`

**Review by stephanosio — APPROVED** [review]
Thanks for your effort implementing this. Glad to see that the Atmel platforms will now have proper pinctrl support.

I have not gone through every single pin mapping due to the enormous size of this PR, but it looks good to me in general.

**Comment by nandojve** [comment]
Hi all,

I think scan code is broke because of this ` WARNING: You are using ScanCode Toolkit version 3.1.1, however the newer version 30.1.0 is available.`

https://github.com/zephyrproject-rtos/docker_scancode/blob/master/Dockerfile#L3

The Docker is using a pretty old version of codescan. I give a try locally using `30.1.0` and than running manually the script without errors. I think we can even update to use Python 3.

**Pull request merged**